contrib/endpoints/src/api_manager/firebase_rules/firebase_request.cc

```diff
@@ -0,0 +1,402 @@
+#include "contrib/endpoints/src/api_manager/firebase_rules/firebase_request.h"
+#include "contrib/endpoints/src/api_manager/utils/marshalling.h"
+#include "contrib/endpoints/src/api_manager/utils/url_util.h"
+#include "google/protobuf/util/message_differencer.h"
+
+#include <algorithm>
+#include <sstream>
+using ::google::api_manager::utils::Status;
+using ::google::api_manager::proto::TestRulesetResponse;
+using ::google::protobuf::util::MessageDifferencer;
+using ::google::protobuf::Map;
+
+namespace google {
+namespace api_manager {
+namespace firebase_rules {
+
+namespace {
+
+const char kToken[] = "token";
+const char kAuth[] = "auth";
+const char kPath[] = "path";
+const char kMethod[] = "method";
+const char kHttpGetMethod[] = "GET";
+const char kHttpPostMethod[] = "POST";
+const char kHttpHeadMethod[] = "HEAD";
+const char kHttpOptionsMethod[] = "OPTIONS";
+const char kHttpDeleteMethod[] = "DELETE";
+const char kFirebaseCreateMethod[] = "create";
+const char kFirebaseGetMethod[] = "get";
+const char kFirebaseDeleteMethod[] = "delete";
+const char kFirebaseUpdateMethod[] = "update";
+const char kV1[] = "/v1";
+const char kTestQuery[] = ":test?alt=json";
+
+void SetProtoValue(const std::string &key,
+                   const ::google::protobuf::Value &value,
+                   ::google::protobuf::Value *head) {
+  ::google::protobuf::Struct *s = head->mutable_struct_value();
+  Map<std::string, google::protobuf::Value> *fields = s->mutable_fields();
+  (*fields)[key] = value;
+}
+
+// Convert HTTP method to Firebase specific method.
+std::string GetOperation(const std::string &httpMethod) {
+  if (httpMethod == kHttpPostMethod) {
+    return kFirebaseCreateMethod;
+  }
+
+  if (httpMethod == kHttpGetMethod || httpMethod == kHttpHeadMethod ||
+      httpMethod == kHttpOptionsMethod) {
+    return kFirebaseGetMethod;
+  }
+
+  if (httpMethod == kHttpDeleteMethod) {
+    return kFirebaseDeleteMethod;
+  }
+
+  return kFirebaseUpdateMethod;
+}
+}
+
+// Constructor
+FirebaseRequest::FirebaseRequest(
+    const std::string &ruleset_name, ApiManagerEnvInterface *env,
+    std::shared_ptr<context::RequestContext> context)
+    : env_(env),
+      context_(context),
+      ruleset_name_(ruleset_name),
+      firebase_server_(
+          context->service_context()->config()->GetFirebaseServer()),
+      current_status_(Status::OK),
+      is_done_(false),
+      next_request_(nullptr) {
+  firebase_http_request_.url =
+      firebase_server_ + kV1 + "/" + ruleset_name + kTestQuery;
+  firebase_http_request_.method = kHttpPostMethod;
+  firebase_http_request_.token_type =
+      auth::ServiceAccountToken::JWT_TOKEN_FOR_FIREBASE;
+  external_http_request_.token_type =
+      auth::ServiceAccountToken::JWT_TOKEN_FOR_AUTHORIZATION_SERVICE;
+
+  // Update the first request to be sent which is the TestRulesetRequest
+  // request.
+  SetStatus(UpdateRulesetRequestBody(RepeatedPtrField<FunctionCall>()));
```

Comment:
Here I am confused now.   It will help if you can add comment about the data flow:
What ESP should do first, what Firebase will response,  and how ESP will handle the response.